ddtrace/internal/ci_visibility/api/_test.py

```diff
@@ -128,29 +153,122 @@ def finish_itr_skipped(self) -> None:
         self.mark_itr_skipped()
         self.finish_test(TestStatus.SKIP)
 
-    def make_early_flake_retry_from_test(self, original_test_id: InternalTestId, retry_number: int) -> None:
-        if self.parent is None:
-            raise ValueError("Cannot make early flake retry from test without a parent")
-
-        new_test_id = InternalTestId(
-            original_test_id.parent_id, original_test_id.name, original_test_id.parameters, retry_number
-        )
-        self.parent.add_child(
-            new_test_id,
-            self.__class__(
-                original_test_id.name,
-                self._session_settings,
-                parameters=self._parameters,
-                codeowners=self._codeowners,
-                source_file_info=self._source_file_info,
-                initial_tags=self._tags,
-                is_early_flake_retry=True,
-            ),
+    def make_early_flake_retry_from_test(self, retry_number: int) -> "TestVisibilityTest":
+        if self._parameters is not None:
+            raise ValueError("Cannot create an early flake retry from a test with parameters")
+        retry_test = self.__class__(
+            self.name,
+            self._session_settings,
+            codeowners=self._codeowners,
+            source_file_info=self._source_file_info,
+            initial_tags=self._tags,
+            is_efd_retry=True,
+            is_new=self._is_new,
         )
+        retry_test.parent = self.parent
+
+        return retry_test
 
     def add_coverage_data(self, coverage_data: Dict[Path, CoverageLines]) -> None:
         self._coverage_data.add_covered_files(coverage_data)
 
     def set_parameters(self, parameters: str) -> None:
         self._parameters = parameters
         self.set_tag(test.PARAMETERS, self._parameters)
+
+    def is_new(self):
+        # NOTE: this is a hack because tests with parameters cannot currently be counted as new (due to EFD design
+        # decisions)
+        return self._is_new and (self._parameters is None)
+
+    # EFD (Early Flake Detection) functionality
+    def efd_should_retry(self):
+        efd_settings = self._session_settings.efd_settings
+        if not efd_settings.enabled:
+            return False
+
+        if self.get_session().efd_is_faulty_session():
+            return False
+
+        if self._efd_abort_reason is not None:
+            return False
+
+        if not self.is_new():
+            return False
+
+        if self._efd_initial_finish_time_ns is None:
+            log.debug("Early Flake Detection: efd_should_retry called but test has initial result")
+            return False
+
+        initial_duration = (self._efd_initial_finish_time_ns - self._span.start_ns) / 1e9
```

Comment:
I think every duration variable should have the unit in the name (e.g., `initial_duration_seconds`), but then again the whole repo is like this. If we are consistent to use units in everything that _isn't_ seconds, that might be good enough, but I don't know if we currently are.